# <span style="color:blue"> 이토록 쉬운 통계 & R </span>

<img src="../image/cover.png" width="200" height="400" align="left">  임경덕 지음

## PART 4 데이터 분석 도구, R

## 90% 순서대로 살펴보는 BR31-4 (33일차)

   * 데이타프로젝트 (http://dataproject.xyz) 또는 (http://cafe.naver.com/dataartproject) 에 미리 저장된 스크립트를 다운 받을 수 있습니다.
   * 스크립트를 활용하면 직접 명령어를 입력하지 않고 저장된 명령어들을 실행할 수 있습니다.
   * 지금 설치된 R 버젼은 3.4.3 입니다. 제목 BR31 은 Basic R 31 로 3.1 버젼입니다. 그러나 큰 차이는 없을 것 같습니다.
   * page 274 subway 예제로 \$ 를 활용한 변수 선택하기 계속


### 데이터 파일은 dataartproject.xyz 에 있습니다.

In [17]:
# subway 데이터 읽기
subway = read.csv("Practice/data/subway.csv")

In [18]:
# 변수명 조정, 여기서는 일력로 붙히는것 때문에 paste 대신에 paste0 을 사용했습니다. 
names(subway) = c("Station", "Date", "InOut", paste0("H", 5:25))

In [19]:
names(subway)

[1] "Station" "Date"    "InOut"   "H5"      "H6"      "H7"      "H8"     
 [8] "H9"      "H10"     "H11"     "H12"     "H13"     "H14"     "H15"    
[15] "H16"     "H17"     "H18"     "H19"     "H20"     "H21"     "H22"    
[22] "H23"     "H24"     "H25"

In [20]:
# 만일 paste 를 사용하려면 sep 를 인자로 주면 됩니다.
names(subway) = c("Station", "Date", "InOut", paste("H", 5:25, sep=""))

In [21]:
names(subway)

[1] "Station" "Date"    "InOut"   "H5"      "H6"      "H7"      "H8"     
 [8] "H9"      "H10"     "H11"     "H12"     "H13"     "H14"     "H15"    
[15] "H16"     "H17"     "H18"     "H19"     "H20"     "H21"     "H22"    
[22] "H23"     "H24"     "H25"

In [22]:
# str() 함수로 데이터 구조 확인하기
str(subway)

'data.frame':	7378 obs. of  24 variables:
 $ Station: Factor w/ 119 levels "가락시장(340)",..: 52 52 52 52 52 52 52 52 52 52 ...
 $ Date   : Factor w/ 31 levels "2016-12-01","2016-12-02",..: 1 1 2 2 3 3 4 4 5 5 ...
 $ InOut  : Factor w/ 2 levels "승차","하차": 1 2 1 2 1 2 1 2 1 2 ...
 $ H5     : num  411 280 434 255 543 260 372 179 499 465 ...
 $ H6     : num  552 1680 520 1609 467 ...
 $ H7     : num  1792 4629 1695 4504 982 ...
 $ H8     : num  3429 10313 3456 10225 2080 ...
 $ H9     : num  2737 5586 2848 5322 3000 ...
 $ H10    : num  2430 2920 2568 3050 3077 ...
 $ H11    : num  3161 2673 3398 2435 4340 ...
 $ H12    : num  3433 2506 3926 3037 5121 ...
 $ H13    : num  3364 2393 3405 3103 4781 ...
 $ H14    : num  2767 2648 2876 2888 4335 ...
 $ H15    : num  3822 2535 5175 3391 6565 ...
 $ H16    : num  4145 2905 5180 3692 8023 ...
 $ H17    : num  4984 3084 6141 3545 6418 ...
 $ H18    : num  9595 3715 10682 5575 6575 ...
 $ H19    : num  5217 2723 6169 3389 4904 ...
 $ H20    : num  34

   * 데이터 형식은 data.frame 입니다.
   * 관측치 갯수는 7378 obs 이고 변수는 24개 입니다.
   * 다음으로 24개 변수의 특징이 있습니다.
   * Station, Date, InOut 은 범주형 변수이며, 각각 119 수준, 31 수준, 2 수준으로 되어 있습니다.

In [23]:
# subway 에서 4번째 변수를 가쟈오기
subway[,4]

[1]  411  280  434  255  543  260  372  179  499  465  489  292  406  240
  [15]  431  298  405  270  433  255  303  169  429  375  430  257  408  300
  [29]  395  295  356  280  466  232  327  157  421  374  427  314  440  279
  [43]  377  286  420  286  481  326  320  175  453  371  414  328  393  281
  [57]  395  298  380  279  429  251   87  186   84  175   87   66  105   52
  [71]   89  192   87  194   78  191   84  183   83  196   85   58   67   36
  [85]   87  193   77  202   87  205   92  183   91  180   85   72   70   41
  [99]   82  208   94  188   80  158   81  172   79  175   79   54  102   72
 [113]   71  163   80  170   80  157   88  176   97  161  119  111  139  210
 [127]  124  194  229   71  236   56  101  198  117  216  113  200  119  190
 [141]  117  200  216   65  188   55  108  200  111  210  140  206  113  186
 [155]  120  187  261   84  243   50  107  198  113  207  128  229  143  202
 [169]  140  178  275   67  198   50  118  188  123  179  123  177  147  194
 [183]  149  176  257   84  199  125  163  100  236   85  257   33  174   98
 [197]  159  120  176   93  158  116  194   98  215   85  222   31  168   91
 [211]  183  107  173   94  162  119  155   74  213   66  248   40  182   93
 [225]  182  111  186   94  140  105  181   90  276   78  240   46  183   98
 [239]  190  103  176   90  178   83  173   77  216   72   85  148   80  145
 [253]   96   92   67   22   61  154   92  160   74  135   94  124   83  127
 [267]   78   81   56   20   67  139   82  153   81  144   74  109   73  144
 [281]   88   85   69   19   60  145   66  141   88  135   77  133   85  137
 [295]   89   82   47   18   62  133   78  136   66  130   86  102   79  137
 [309]   89   70  616   78  592   82  618   56  152   21  620   98  574   98
 [323]  578   90  584   76  588   86  560   75  140   21  632   91  629   96
 [337]  628   88  634   78  588   63  596   57  150   29  625   81  622   76
 [351]  648   89  571   80  578   78  604   63  128   19  618   96  653   98
 [365]  571  100  586  102  393   95  188   48  289   91  306  101  253   66
 [379]  141   47  319   94  292   88  298   92  305   92  314   95  268   60
 [393]  147   32  334   92  308   93  312   90  286   64  271   72  259   60
 [407]  153   40  286   78  317   93  302   74  236   73  294   58  261   63
 [421]  127   25  307   80  303   89  284   79  286   73  271   75  213   52
 [435]  153   98  133   97  153   80   68   31  184   91  164   89  155   86
 [449]  151   74  153   84  130   68   78   37  176   88  168  100  168   86
 [463]  165   85  164   89  146   82   84   43  179   95  160   90  149   84
 [477]  127   76  151   85  123   75   76   28  152   81  151   80  138   77
 [491]  141   79  152   84  121   75  499  129  518  126  464  139  239   62
 [505]  558  141  494  133  515  116  480  154  504  137  442  127  232   55
 [519]  553  129  538  145  510  138  489  134  437  114  434  153  265   72
 [533]  523  137  513  131  476  135  418  114  480  118  428  144  227   63
 [547]  518  146  466  119  414  128  443  123  415  123  328  119  120   27
 [561]  118   27  115   28   42   27  126   29  124   28  119   30  116   32
 [575]   98   33  124   31   48   21  103   31  130   34  109   30  101   30
 [589]  105   35  120   32   39   23   99   28  127   29   98   30  107   27
 [603]   94   28   81   23   48   15  124   30  112   22   93   24   83   24
 [617]   90   29   63   23   49   83   51   68   59   40   59   18   56   74
 [631]   50   68   76   66   65   67   67   61   60   37   60   15   54   69
 [645]   57   70   57   68   67   67   54   64   65   32   43   17   50   75
 [659]   54   71   64   74   57   72   67   74   65   55   48   33   44   73
 [673]   56   77   58   65   54   72  113  121   53   39  137  122  144  135
 [687]  128   93   86   87  135  118  119  107  134  119  133  100  145  111
 [701]  142  105   73   76  126   90  156  129  127  106  149  132  131  107
 [715]  139   96   99   76  127  100  135  100  148  109  134  106  149  146

In [24]:
# 같은 방법을 $ 를 활용해서
subway$H5

[1]  411  280  434  255  543  260  372  179  499  465  489  292  406  240
  [15]  431  298  405  270  433  255  303  169  429  375  430  257  408  300
  [29]  395  295  356  280  466  232  327  157  421  374  427  314  440  279
  [43]  377  286  420  286  481  326  320  175  453  371  414  328  393  281
  [57]  395  298  380  279  429  251   87  186   84  175   87   66  105   52
  [71]   89  192   87  194   78  191   84  183   83  196   85   58   67   36
  [85]   87  193   77  202   87  205   92  183   91  180   85   72   70   41
  [99]   82  208   94  188   80  158   81  172   79  175   79   54  102   72
 [113]   71  163   80  170   80  157   88  176   97  161  119  111  139  210
 [127]  124  194  229   71  236   56  101  198  117  216  113  200  119  190
 [141]  117  200  216   65  188   55  108  200  111  210  140  206  113  186
 [155]  120  187  261   84  243   50  107  198  113  207  128  229  143  202
 [169]  140  178  275   67  198   50  118  188  123  179  123  177  147  194
 [183]  149  176  257   84  199  125  163  100  236   85  257   33  174   98
 [197]  159  120  176   93  158  116  194   98  215   85  222   31  168   91
 [211]  183  107  173   94  162  119  155   74  213   66  248   40  182   93
 [225]  182  111  186   94  140  105  181   90  276   78  240   46  183   98
 [239]  190  103  176   90  178   83  173   77  216   72   85  148   80  145
 [253]   96   92   67   22   61  154   92  160   74  135   94  124   83  127
 [267]   78   81   56   20   67  139   82  153   81  144   74  109   73  144
 [281]   88   85   69   19   60  145   66  141   88  135   77  133   85  137
 [295]   89   82   47   18   62  133   78  136   66  130   86  102   79  137
 [309]   89   70  616   78  592   82  618   56  152   21  620   98  574   98
 [323]  578   90  584   76  588   86  560   75  140   21  632   91  629   96
 [337]  628   88  634   78  588   63  596   57  150   29  625   81  622   76
 [351]  648   89  571   80  578   78  604   63  128   19  618   96  653   98
 [365]  571  100  586  102  393   95  188   48  289   91  306  101  253   66
 [379]  141   47  319   94  292   88  298   92  305   92  314   95  268   60
 [393]  147   32  334   92  308   93  312   90  286   64  271   72  259   60
 [407]  153   40  286   78  317   93  302   74  236   73  294   58  261   63
 [421]  127   25  307   80  303   89  284   79  286   73  271   75  213   52
 [435]  153   98  133   97  153   80   68   31  184   91  164   89  155   86
 [449]  151   74  153   84  130   68   78   37  176   88  168  100  168   86
 [463]  165   85  164   89  146   82   84   43  179   95  160   90  149   84
 [477]  127   76  151   85  123   75   76   28  152   81  151   80  138   77
 [491]  141   79  152   84  121   75  499  129  518  126  464  139  239   62
 [505]  558  141  494  133  515  116  480  154  504  137  442  127  232   55
 [519]  553  129  538  145  510  138  489  134  437  114  434  153  265   72
 [533]  523  137  513  131  476  135  418  114  480  118  428  144  227   63
 [547]  518  146  466  119  414  128  443  123  415  123  328  119  120   27
 [561]  118   27  115   28   42   27  126   29  124   28  119   30  116   32
 [575]   98   33  124   31   48   21  103   31  130   34  109   30  101   30
 [589]  105   35  120   32   39   23   99   28  127   29   98   30  107   27
 [603]   94   28   81   23   48   15  124   30  112   22   93   24   83   24
 [617]   90   29   63   23   49   83   51   68   59   40   59   18   56   74
 [631]   50   68   76   66   65   67   67   61   60   37   60   15   54   69
 [645]   57   70   57   68   67   67   54   64   65   32   43   17   50   75
 [659]   54   71   64   74   57   72   67   74   65   55   48   33   44   73
 [673]   56   77   58   65   54   72  113  121   53   39  137  122  144  135
 [687]  128   93   86   87  135  118  119  107  134  119  133  100  145  111
 [701]  142  105   73   76  126   90  156  129  127  106  149  132  131  107
 [715]  139   96   99   76  127  100  135  100  148  109  134  106  149  146

## factor() 함수로 범주형 변수 만들기

범주형 변수의 수준은 일반적으로 문자 형식으로 되어 있습니다. 이것을 숫자형으로 표시하면 처리 하기가 더 간편하게 됩니다. 
   * 예를 들어 성별 을 표시할때, 남자, 여자 보다는 1은 남자, 2는 여자로 코드집화하여 데이터에는 1, 2 로 표시하는 것이 편합니다.
   * 그러나 R 에서는 숫자로 표시되면 숫자 형식으로 인식하기 때문에 문제가 있을수 있습니다.
   * 이 경우 factor() 함수를 이용하여 숫자가 아닌 범주형 변수 라는 것을 R 에 인식시켜향 합니다.

In [25]:
demo = read.csv("Practice/data/gender_age.csv")

In [26]:
demo

Gender Age
1 1      30 
2 2      20 
3 1      20 
4 2      20 
5 1      30 
6 2      30

In [27]:
# 데이터의 구조를 파악합니다.
str(demo)

'data.frame':	6 obs. of  2 variables:
 $ Gender: int  1 2 1 2 1 2
 $ Age   : int  30 20 20 20 30 30


   * Gender 를 숫자로 인식하고 있습니다. 이것을 범주형 변수로 인식하게 하여야 합니다.

In [28]:
demo$Gender

[1] 1 2 1 2 1 2

In [29]:
# 범주형 변수로 인식
factor(demo$Gender)

[1] 1 2 1 2 1 2
Levels: 1 2

In [30]:
str(factor(demo$Gender))

 Factor w/ 2 levels "1","2": 1 2 1 2 1 2


### $ 를 활용한 변수 수정과 추가하기

In [33]:
# Gender 를 변수형 변수로 인식
demo$Gender = factor(demo$Gender)

In [34]:
str(demo)

'data.frame':	6 obs. of  2 variables:
 $ Gender: Factor w/ 2 levels "1","2": 1 2 1 2 1 2
 $ Age   : int  30 20 20 20 30 30


In [35]:
# 데이터 프레임에 변수 추가 하기
paste0(demo$Gender,"_",demo$Age,"대")

[1] "1_30대" "2_20대" "1_20대" "2_20대" "1_30대" "2_30대"

In [36]:
# 이것을 새로운 변수 "Group" 로 추가합니다.
demo$Group = paste0(demo$Gender,"_",demo$Age,"대")

In [37]:
# demo 의 구조를 보면
str(demo)

'data.frame':	6 obs. of  3 variables:
 $ Gender: Factor w/ 2 levels "1","2": 1 2 1 2 1 2
 $ Age   : int  30 20 20 20 30 30
 $ Group : chr  "1_30대" "2_20대" "1_20대" "2_20대" ...
